From 90c67432915c58f491edcefae6f2f087693d2b6c Mon Sep 17 00:00:00 2001
From: Ryan O'Leary <ryanoleary@google.com>
Date: Fri, 25 Jun 2021 13:46:53 -0700
Subject: [PATCH] Change FLASH_BASE to 0xFFF80000

---
 QemuFspPkg/QemuFspPkg.fdf | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/QemuFspPkg/QemuFspPkg.fdf b/QemuFspPkg/QemuFspPkg.fdf
index b08e3c2a40..846764ec10 100644
--- a/QemuFspPkg/QemuFspPkg.fdf
+++ b/QemuFspPkg/QemuFspPkg.fdf
@@ -14,7 +14,7 @@
 ##
 
 [Defines]
-DEFINE FLASH_BASE           = 0xFFFC8000
+DEFINE FLASH_BASE           = 0xFFF80000
 DEFINE FLASH_BLOCK_SIZE     = 0x00000100
 
 #
-- 
2.32.0.93.g670b81a890-goog

